assembly/src/todo/mod.rs

```diff
@@ -0,0 +1,432 @@
+        self.cb_table
+    }
+
+    pub fn compile<S>(&mut self, source: S) -> Program
+    where
+        S: AsRef<str>,
+    {
+        let source = source.as_ref();
+        let ast = ProgramAst::try_from(source).unwrap();
+
+        self.compile_ast(&ast)
+    }
+
+    pub fn compile_ast(&mut self, ast: &ProgramAst) -> Program {
+        let ProgramAst { local_procs, body } = ast;
+
+        self.locals.clear();
+        self.blocks.clear();
+        self.spans.clear();
+
+        let root = self.compile_body(body.iter());
+
+        Program::new(root)
+    }
+
+    fn start_span<I, T>(&mut self, iter: &I)
+    where
+        I: Iterator<Item = T>,
+    {
+        self.end_span();
+        self.spans.push(vec_with_iter_capacity(iter));
+    }
+
+    fn append_to_span(&mut self, operation: Operation) {
+        if let Some(span) = self.spans.last_mut() {
+            span.push(operation);
+        }
+    }
+
+    fn end_span(&mut self) {
+        self.spans
+            .pop()
+            .filter(|v| !v.is_empty())
+            .map(CodeBlock::new_span)
+            .map(|span| self.blocks.last_mut().map(|block| block.push(span)));
+    }
+
+    fn start_block<I, T>(&mut self, iter: &I)
+    where
+        I: Iterator<Item = T>,
+    {
+        self.blocks.push(vec_with_iter_capacity(iter));
+    }
+
+    fn pop_block(&mut self) -> Option<CodeBlock> {
+        self.blocks
+            .pop()
+            .filter(|v| !v.is_empty())
+            .map(parsers::combine_blocks)
+    }
+
+    fn append_block(&mut self, block: CodeBlock) {
+        self.blocks.last_mut().map(|blocks| blocks.push(block));
+    }
+
+    fn compile_body<A, N>(&mut self, ast: A) -> CodeBlock
+    where
+        A: Iterator<Item = N>,
+        N: Borrow<Node>,
+    {
+        self.start_block(&ast);
+        self.start_span(&ast);
+
+        for node in ast {
+            match node.borrow() {
+                Node::Instruction(instruction) => {
+                    self.compile_instruction(instruction);
+                }
+
+                Node::IfElse(t, f) => {
+                    let t = self.compile_body(t.iter());
+                    let f = self.compile_body(f.iter());
+
+                    let block = CodeBlock::new_split(t, f);
+
+                    self.append_block(block);
+                }
+
+                Node::Repeat(n, nodes) => {
+                    let block = self.compile_body(nodes.iter());
+
+                    for _ in 0..*n {
+                        self.append_block(block.clone());
+                    }
+                }
+
+                Node::While(nodes) => {
+                    let block = self.compile_body(nodes.iter());
+                    let block = CodeBlock::new_loop(block);
+
+                    self.append_block(block);
+                }
+            }
+        }
+
+        self.end_span();
+        self.pop_block().unwrap()
+    }
+
+    fn compile_instruction(&mut self, instruction: &Instruction) {
```

Comment:
This should take `span_ops` as a mutable vector. In general, the signature of this function should be something like:
```Rust
fn parse_op_token(
    &self,
    instruction: &Instruction,
    span_ops: &mut Vec<Operation>,
    num_proc_locals: u32,
    decorators: &mut DecoratorList,
    in_kernel: bool,
) -> Result<(), AssemblyError> {
```
But we can add `num_proc_locals`, `decorators`, and `in_kernel` later.

Reply:
SpanBuilder created; receiving also context